File: packages/android-client/FishjamClient/src/main/java/com/fishjamdev/client/media/Track.kt
package com.fishjamdev.client.media

import com.fishjamdev.client.models.Metadata
import org.webrtc.MediaStreamTrack
import java.util.UUID

open class Track(
  internal val mediaTrack: MediaStreamTrack?,
  val endpointId: String,
  private var rtcEngineId: String?,
  metadata: Metadata = mapOf(),
  /**
   * Every track can have 2 ids:
   * - the one from rtc engine in the form of <peerid>:<trackid>
   * - the one from webrtc library in the form of uuidv4 string (different than one from the engine)
   * It's always confusing when to use which one and we need to keep some kind of mapping.
   * What's worse a track might be sometimes missing one of the ids
   * - rtc engine id is missing when a local track is created without established connection
   * - webrtc id is missing when we get a track from rtc engine but is not yet negotiated
   * So we have a third id that is created by our client. It's always there, it's never changing
   * and we're always using it unless we talk to rtc engine or webrtc. The user sees just this id,
   * unless they want to debug something.
   */
  private val id: String = UUID.randomUUID().toString()
) {
  fun id(): String = id

  internal fun webrtcId(): String = mediaTrack?.id() ?: ""

  var metadata: Metadata = metadata
    internal set

  fun isEnabled(): Boolean = mediaTrack?.enabled() ?: false

  fun setEnabled(isEnabled: Boolean) {
    mediaTrack?.setEnabled(isEnabled)
  }

  internal fun setRTCEngineId(id: String) {
    rtcEngineId = id
  }

  internal fun getRTCEngineId(): String? = rtcEngineId
}

File: packages/ios-client/Sources/MembraneRTC/Media/Tracks/Track.swift
//DONE

import WebRTC

open class Track {
    internal var mediaTrack: RTCMediaStreamTrack?
    let endpointId: String
    internal var rtcEngineId: String?
    var metadata: Metadata
    public let id: String

    /**
    * Every track can have two IDs:
    * - the one from RTC engine in the format <peerid>:<trackid>
    * - the one from WebRTC library in the form of a UUID v4 string (different than the one from the engine)
    * It's often confusing when to use which one, and a mapping is typically needed.
    * Moreover, a track may sometimes be missing one of the IDs:
    * - RTC engine ID is missing when a local track is created without an established connection
    * - WebRTC ID is missing when a track is received from an RTC engine but has not been negotiated
    * As a solution, there is a third ID created by our client. It's always present, never changes,
    * and is always used unless communicating specifically with the RTC engine or WebRTC. This ID is what
    * the user sees, unless they need to debug something.
    */

    init(
        mediaTrack: RTCMediaStreamTrack?, endpointId: String, rtcEngineId: String?, metadata: Metadata = Metadata(),
        id: String = UUID().uuidString
    ) {
        self.mediaTrack = mediaTrack
        self.endpointId = endpointId
        self.rtcEngineId = rtcEngineId
        self.metadata = metadata
        self.id = id
    }

    internal var webrtcId: String {
        return mediaTrack?.trackId ?? ""
    }

    internal var enabled: Bool {
        get {
            return mediaTrack?.isEnabled ?? false
        }
        set {
            mediaTrack?.isEnabled = newValue
        }
    }

}
